File: src/voicetype/glossary.py
"""Glossary-based transcript correction."""

import re

from voicetype.config import GlossaryEntry

# Cache the compiled regex + replacement map. The key is a tuple of
# (source, replacement) pairs; it changes only when settings are saved.
# The list object identity works because config.glossary is replaced wholesale
# on load/save, but a content-derived key is safer for mutations in place.
_glossary_cache: dict[tuple[tuple[str, str], ...], tuple[re.Pattern, dict[str, str]]] = {}


def _build_glossary_cache(entries: list[GlossaryEntry]) -> tuple[re.Pattern, dict[str, str]] | None:
    """Build (or retrieve from cache) the compiled regex and replacement map."""
    valid = [
        (entry.source.strip(), entry.replacement.strip())
        for entry in entries
        if entry.source.strip() and entry.replacement.strip()
    ]
    if not valid:
        return None

    key = tuple(sorted(valid, key=lambda item: len(item[0]), reverse=True))
    cached = _glossary_cache.get(key)
    if cached is not None:
        return cached

    # Keep only the first occurrence of each source when duplicates exist.
    seen: set[str] = set()
    unique: list[tuple[str, str]] = []
    for source, replacement in key:
        if source not in seen:
            seen.add(source)
            unique.append((source, replacement))

    pattern = re.compile("|".join(re.escape(source) for source, _ in unique))
    replacement_map = {source: replacement for source, replacement in unique}

    # Bound cache size — evict if it grows too large (unlikely in practice).
    if len(_glossary_cache) > 16:
        _glossary_cache.clear()

    result = (pattern, replacement_map)
    _glossary_cache[key] = result
    return result


def invalidate_glossary_cache() -> None:
    """Drop the compiled glossary cache (call after settings change)."""
    _glossary_cache.clear()


def apply_glossary(text: str, entries: list[GlossaryEntry]) -> str:
    """Apply user-defined term replacements to recognized text."""
    if not text or not entries:
        return text

    cached = _build_glossary_cache(entries)
    if cached is None:
        return text

    pattern, replacement_map = cached
    return pattern.sub(lambda m: replacement_map[m.group(0)], text)

File: desktop/src/main/services/glossary.ts
// Glossary-based transcript correction — port of voicetype/glossary.py.
// All sources are compiled into one alternation regex (longest first) and
// replaced in a single pass; compiled patterns are content-keyed and cached.

import type { GlossaryEntry } from '../../shared/types'

interface CompiledGlossary {
  pattern: RegExp
  replacementMap: Map<string, string>
}

const glossaryCache = new Map<string, CompiledGlossary>()

function normalize(entries: GlossaryEntry[]): Array<[string, string]> {
  return entries
    .map((e) => [e.source.trim(), e.replacement.trim()] as [string, string])
    .filter(([source, replacement]) => source && replacement)
    .sort((a, b) => b[0].length - a[0].length)
}

function buildGlossary(entries: GlossaryEntry[]): CompiledGlossary | null {
  const valid = normalize(entries)
  if (!valid.length) return null

  const key = JSON.stringify(valid)
  const cached = glossaryCache.get(key)
  if (cached) return cached

  const seen = new Set<string>()
  const unique: Array<[string, string]> = []
  for (const [source, replacement] of valid) {
    if (!seen.has(source)) {
      seen.add(source)
      unique.push([source, replacement])
    }
  }

  const pattern = new RegExp(unique.map(([source]) => source.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|'), 'gu')
  const replacementMap = new Map(unique)

  if (glossaryCache.size >= 16) {
    glossaryCache.delete(glossaryCache.keys().next().value as string)
  }
  const compiled = { pattern, replacementMap }
  glossaryCache.set(key, compiled)
  return compiled
}

export function invalidateGlossaryCache(): void {
  glossaryCache.clear()
}

export function applyGlossary(text: string, entries: GlossaryEntry[]): string {
  if (!text || !entries.length) return text
  const compiled = buildGlossary(entries)
  if (!compiled) return text
  return text.replace(compiled.pattern, (m) => compiled.replacementMap.get(m) ?? m)
}
